psydac/feec/derivatives.py

```diff
@@ -1,121 +1,248 @@
 # -*- coding: UTF-8 -*-
 
 import numpy as np
-from numpy import zeros
-from numpy import concatenate
-from numpy import block
 
-from scipy import kron
-from scipy.sparse import identity
-from scipy.sparse import csr_matrix
+from psydac.linalg.stencil import StencilMatrix, StencilVectorSpace
+from psydac.linalg.kron    import KroneckerStencilMatrix as Stencil_kron
+from psydac.linalg.block   import ProductSpace, BlockVector, BlockLinearOperator, BlockMatrix
+from psydac.fem.vector     import ProductFemSpace
 
-def d_matrix(n):
+def d_matrix(n, p, P):
     """creates a 1d incidence matrix.
     The final matrix will have a shape of (n,n-1)
 
     n: int
         number of nodes
+        
+    p : int
+        pads
+        
+    P : bool
+        periodicity
     """
-    M = zeros((n,n))
-    for i in range(0, n):
-        M[i,i] = 1.
-        if i>0:
-            M[i,i-1] = -1.
-    return csr_matrix(M[1:n,:])
+    
+    V1 = StencilVectorSpace([n], [p], [P])
+    V2 = StencilVectorSpace([n-1], [p], [P])
+    M = StencilMatrix(V1, V2)
+    
+
+    for i in range(n):
+        M._data[p+i,p] = -1.
+        M._data[p+i,p+1] = 1.
+    return M
+    
+def identity(n, p, P):
```

Comment:
Why do we store an identity matrix in stencil form, rather than creating an `IdentityMatrix` class (or `IdentityLinearOperator`)? An object of such a class should require no storage at all (apart from some metadata).

Reply:
yes we could do that it's ,this is just a first implementation but yes i agree we should use something like `IdentityMatrix`